keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1289 @@
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: silver
+parameters:
+  iops: "500"
+  throughput: "50MiB/s"
+```
+
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gcepd
+provisioner: pd.csi.storage.gke.io
+volumeBindingMode: WaitForFirstConsumer
+```
+
+The driver is responsible for parsing and validating these parameters and making sure duplicate configuration is not allowed. End users will specify  the VolumeAttributesClass in the PVC object:
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  VolumeAttributesClassName: silver
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+#### Volume Attributes Updates
+
+A customer monitors the performance of a Cassandra cluster. They notice that volume IO is causing a bottleneck in performance. They update the PVCs in question by changing to a higher tier VolumeAttributesClass managed by the cluster administrator to increase the provisioned IO and realize improved performance.
+
+Before:
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: silver
+parameters:
+  iops: "500"
+  throughput: "50MiB/s"
+```
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  VolumeAttributesClassName: silver
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+After:
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: gold
+parameters:
+  iops: "1000"
+  throughput: "100MiB/s"
+```
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test-pv-claim
+spec:
+  storageClassName: csi-sc-example
+  VolumeAttributesClassName: gold
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+```
+
+#### Administrator Quota Restrictions
+
+A cluster admin wants to control costs while giving application developers the freedom to optimize their applications. They set a per VolumeAttributesClass limit to the maximum count of PVCs that can be specified in a cluster ResourceQuota. When an application dev modifies a PVC to request a higher tier of VolumeAttributesClass but there is no quota, the request is rejected. As the ResourceQuota is a cluster-scoped object, only the cluster admin and not application devs can change limits.
+
+An example of define ResourceQuota for VolumeAttributesClass:
+
+
+```
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  name: vacquota
+spec:
+  hard:
+  // Across all persistent volume claims associated with the 
+  // <volume-attributes-class-name>, the total number of persistent volume claims 
+  // that can exist in the namespace.
+<volume-attributes-class-name>.VolumeAttributesClass.storage.k8s.io/persistentvolumeclaims: "5" 
+  ...
+```
+
```

Comment:
Did we document the case where what happens if an admin modifies parameters of `VolumeAttribtueClass` ? I didn't see mention of the use case anywhere.

Reply:
The above example in line 342

```
apiVersion: storage.k8s.io/v1alpha1
kind: VolumeAttributesClass
metadata:
  name: gold
parameters:
  iops: "1000"
  throughput: "100MiB/s"
```

Reply:
Oh i think you were mentioning if the parameters are mutable. The parameters are immutable in the VAC but it is mutable in CSI spec.